torch/onnx/_internal/fx/passes/functionalization.py

```diff
@@ -107,6 +109,7 @@ def _run(self, *args) -> torch.fx.GraphModule:
             decomposition_table={},
             tracing_mode=fx_mode,
             _allow_non_fake_inputs=True,
+            _allow_fake_constant=self.fake_mode is not None,
```

Comment:
As long as the user fakefied the input/model themselves, `fake_mode` will be needed by the `torch._dynamo.export` call, so passing `fake_mode` is needed

Setting `True` regardless of `fake_mode` saves plumbing for the passes, only. Not sure what happens if the user doesn't fakefy the input/model and `True` is passed to `make_fx(_allow_fake_constant=True)`. I traied to keep all call sites aligned: when fake_mode is used, both `torch._dynamo.export` and `make_fx` will leverage them. When `fake_mode` is not in the picture, both `torch._dynamo.export` and `make_fx` ignore them

Reply:
> Not sure what happens if the user doesn't fakefy the input/model and True is passed to make_fx(_allow_fake_constant=True).

Not raising when spotting fake constant. 

I don't like `self.fake_mode` it is too ambiguous. A `fake_mode` is always created by dynamo regardless.

Reply:
Dynamo does not always create `fake_mode`. 
Pytorch PR #100017 recycles user's `fake_mode` when one is specified by `torch._dynamo.export`.
This is the name adopted by PyTorch on several APIs (see `torch._dynamo.export). This PR reuses the same name for the same purpose to avoid confusion.

Reply:
LOL typed too quick I meant "A fake_mode always exists after dynamo call regardless." + (so it doesn't necessarily mean this is the one created by the user, and not the one created by dynamo). 

The two potential source of creation is exactly why in this context `fake_mode` is ambiguous.

Reply:
This is not ambiguity; this is a design decision approved by Edward for the `torch._dynamo.export` API which we are reuising here as a frontend to `torch._dynamo.export`.

If user specifies their `fake_mode`, it is adopted by dynamo. Otherwise, dynamo creates one itself. It is not possible to allow user to fakefy input/model using their `fake_mode` and have `torch._dynamo.export` to create another `fake_mode`. Each instance will have mismatching metadata, which will be detected by dynamo and a `mixed fake mode not yet supported` will be raised.

In feature PRs we can experiment to take non-fake input and model and fakeify internally, but would assume the GPU will have memory to hold input and model without OOM. I have filed a [separate task](https://github.com/pytorch/pytorch/issues/104144) for this to not block an already working solution

Reply:
Should've called offline regarding fake mode ambiguity ... we are still not talking about the same thing.

> but would assume the GPU will have memory to hold input and model without OOM.

This is the topic I'd like for more discussion in the design doc, namely in the e2e scenario and apis we are offering. Right now both model loading and serialization are not handled by us, and both steps naively requires loading entire model into memory and potentially incur OOM.